### PR TITLE
Incorporate JsonPath into exception messages

### PR DESCRIPTION
**Pull request by qwwdfsad**
While figuring out what to do with polymorphism I've come to the conclusion that #1839 have to be done, otherwise the change is way too widespread and intrusive while being really hard to maintain, so I've reverted this work halfway through and stopped on this

### CODE REVIEW

**Comment by qwwdfsad** [comment]
The latest commit is kinda controversial, it seems like we have to make #1266 finally public for both usage and construction

File: formats/json/commonTest/src/kotlinx/serialization/JsonPathTest.kt [hunk 1]
```diff
@@ -32,6 +32,13 @@ class JsonPathTest : JsonTestBase() {
         expectPath("$.i.d[1]") { Json.decodeFromString<Outer>("""{"a":42, "i":{"d":{1:{}}""") }
     }
 
+    @Test
+    fun testUnknownKeyIsProperlyReported() {
+        expectPath("$.i") { Json.decodeFromString<Outer>("""{"a":42, "i":{"foo":42}""") }
+        expectPath("$") { Json.decodeFromString<Outer>("""{"x":{}, "a": 42}""") }
+//        Json.decodeFromString<Outer>("""{"a":42, "x":{}}""")
```
Comment (qwwdfsad) [inline formats/json/commonTest/src/kotlinx/serialization/JsonPathTest.kt:39]: This is the only place where misattribution happens, but I'm not sure how to fix it without a huge performance impact, so maybe we can live with that

File: formats/json/commonTest/src/kotlinx/serialization/JsonPathTest.kt [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+
+    @Test
+    fun testMalformedMap() {
+        expectPath("$.i.d\n") { Json.decodeFromString<Outer>("""{"a":42, "i":{ "d": []""") }
+        expectPath("$\n") { Json.decodeFromString<Map<Int, String>>("""[]""") }
+    }
+
+    @Test
+    fun testMapValue() {
+        expectPath("$.i.d[42]\n") { Json.decodeFromString<Outer>("""{"a":42, "i":{ "d": {42: {"xx":"bar"}}""") }
```
Comment (JakeWharton) [inline formats/json/commonTest/src/kotlinx/serialization/JsonPathTest.kt:79]: Shouldn't these be quoted such as `$.i.d['42']` to indicate object keys and not array indexing? As of now it is indistinguishable (and thus ambiguous) from the 43rd element of an array.

File: formats/json/commonTest/src/kotlinx/serialization/JsonPathTest.kt (hunk 1 above)
Comment (sandwwraith) [reply formats/json/commonTest/src/kotlinx/serialization/JsonPathTest.kt:39]: Please add this comment directly to the code

**Review by sandwwraith — APPROVED** [review]

**Pull request merged**